### PR TITLE
Votenet api

### PR DESCRIPTION
**Pull request by tchaton**

### CODE REVIEW

File: torch_points3d/applications/votenet.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        """Unpack input data from the dataloader and perform necessary pre-processing steps.
+        Parameters:
+            input: a dictionary that contains the data itself and its metadata information.
+        """
+        # Forward through backbone model
+        if self._kpconv_backbone:
+            data = AddOnes()(data)
+            if data.x is not None:
+                data.x = torch.cat([data.x, data.ones.float()], dim=-1)
+            else:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply torch_points3d/applications/votenet.py:240]: I can't install Minkowski Engine on Mac OS without the docker container and I have some trouble with it. So I can't do it. Feel free to add it. Should be pretty simple

**Comment by tchaton** [comment]
I had to remove this tests which had been failing couples of time self.assertTrue(plane_filter(data2).item()) TODO THIS TEST SEEM BROKEN

File: torch_points3d/modules/VoteNet/proposal_module.py [hunk 2]
```diff
@@ -33,20 +34,30 @@ def __init__(
...
         # Object proposal/detection
         # Objectness scores (2), center residual (3),
         # heading class+residual (num_heading_bin*2), size class+residual(num_size_cluster*4)
-        self.conv1 = torch.nn.Conv1d(128, 128, 1)
-        self.conv2 = torch.nn.Conv1d(128, 128, 1)
-        self.conv3 = torch.nn.Conv1d(128, 2 + 3 + num_heading_bin * 2 + self.num_size_cluster * 4 + self.num_class, 1)
-        self.bn1 = torch.nn.BatchNorm1d(128)
-        self.bn2 = torch.nn.BatchNorm1d(128)
+        self.conv_type = conv_type
+        if self.conv_type == "DENSE":
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/proposal_module.py:41]: Same as for the loss. At that point, everything is dense so no need to add complexity until we actually support more than dense

File: torch_points3d/modules/VoteNet/voting_module.py [hunk 3]
```diff
@@ -26,11 +26,19 @@ def __init__(self, vote_factor, seed_feature_dim):
         self.vote_factor = vote_factor
         self.in_dim = seed_feature_dim
         self.out_dim = self.in_dim  # due to residual feature, in_dim has to be == out_dim
-        self.conv1 = torch.nn.Conv1d(self.in_dim, self.in_dim, 1)
-        self.conv2 = torch.nn.Conv1d(self.in_dim, self.in_dim, 1)
-        self.conv3 = torch.nn.Conv1d(self.in_dim, (3 + self.out_dim) * self.vote_factor, 1)
-        self.bn1 = torch.nn.BatchNorm1d(self.in_dim)
-        self.bn2 = torch.nn.BatchNorm1d(self.in_dim)
+        self._conv_type = conv_type
+        if conv_type == "DENSE":
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/voting_module.py:30]: Same, only dense tensor should arrive there at the moment. Let's take the generalisation to non dense formats as a separate piece of work

Comment (tchaton) [reply torch_points3d/modules/VoteNet/voting_module.py:30]: I left the dense here if we want to use it with partial dense for later, if we want to perform clustering on partial dense format.

File: torch_points3d/modules/VoteNet/loss_helper.py [hunk 4]
```diff
@@ -47,11 +47,14 @@ def compute_vote_loss(input, output: VoteNetResults):
...
     # vote_label: Use gather to select B,num_seed,9 from B,num_point,9
     #   with inds in shape B,num_seed,9 and 9 = GT_VOTE_FACTOR * 3
-    seed_gt_votes_mask = torch.gather(input["vote_label_mask"], 1, seed_inds)
-    seed_inds_expand = seed_inds.view(batch_size, num_seed, 1).repeat(1, 1, 3 * GT_VOTE_FACTOR)
-
-    seed_gt_votes = torch.gather(input["vote_label"], 1, seed_inds_expand)
-    seed_gt_votes += output["seed_pos"].repeat(1, 1, 3)
+    if seed_inds.dim() == 1:
+        seed_gt_votes_mask = torch.gather(input["vote_label_mask"], 0, seed_inds).view((batch_size, -1))
+        seed_gt_votes = torch.gather(input["vote_label"], 0, seed_inds.unsqueeze(-1).repeat(1, 3 * GT_VOTE_FACTOR))
```
[the remark this answers is not in the document]
Comment (tchaton) [reply torch_points3d/modules/VoteNet/loss_helper.py:52]: Done !

File: README.md [hunk 5]
```diff
@@ -58,15 +59,16 @@ where each folder contains the dataset related to each task.
...
-* **[MinkowskiEngine](https://github.com/nicolas-chaulet/torch-points3d/tree/master/torch_points3d/modules/MinkowskiEngine)** from Christopher Choy *et al.*: [4D Spatio-Temporal ConvNets: Minkowski Convolutional Neural Networks](https://arxiv.org/abs/1904.08755) (CVPR19)
-* **[VoteNet](https://github.com/nicolas-chaulet/torch-points3d/tree/master/torch_points3d/models/object_detection/votenet.py)** from Charles R. Qi *et al.*: [Deep Hough Voting for 3D Object Detection in Point Clouds](https://arxiv.org/abs/1904.09664) (ICCV 19)
-- **[FCGF](https://github.com/chrischoy/FCGF)** from Christopher Choy _et al._: [Fully Convolutional Geometric Features](https://node1.chrischoy.org/data/publications/fcgf/fcgf.pdf) (ICCV'19)
+- **[PointNet](https://github.com/nicolas-chaulet/torch-points3d/blob/master/torch_points3d/modules/PointNet/modules.py#L54)** from Charles R. Qi _et al._: [PointNet: Deep Learning on Point Sets for 3D Classification and Segmentation](https://arxiv.org/abs/1612.00593) (CVPR 2017)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply README.md]: I copied master directly. No idea why he is showing difference

File: torch_points3d/applications/votenet.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+CUR_FILE = os.path.realpath(__file__)
+DIR_PATH = os.path.dirname(os.path.realpath(__file__))
+PATH_TO_CONFIG = os.path.join(DIR_PATH, "conf/votenet")
+
+log = logging.getLogger(__name__)
+
+MAPPING_BACKBONES_TO_API_NAMES = {"kpconv": "KPConv", "pointnet2": "PointNet2", "rsconv": "RSConv"}
+
+
+def resolve_model(model_config, num_features, kwargs):
```
Comment (nicolas-chaulet) [inline torch_points3d/applications/votenet.py:25]: Why did you create a new method as opposed to using the existing one?

File: torch_points3d/applications/votenet.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        constants.update(dict(model_config.define_constants))
+        define_constants = model_config.define_constants
+        for key in define_constants.keys():
+            value = kwargs.get(key)
+            if value:
+                constants[key] = value
+    resolve(model_config, constants)
+
+
+def sample_to_dense_format(data, num_batches, conv_type, num_points_to_sample):
```
Comment (nicolas-chaulet) [inline torch_points3d/applications/votenet.py:45]: Please add a test for it

File: torch_points3d/applications/votenet.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        self._kwargs = kwargs
+        self._compute_loss = compute_loss
+        self._backbone = backbone
+
+        # 1 - CREATE BACKBONE MODEL
+        if backbone is None:
+            backbone_option = option.backbone
+            backbone_cls = getattr(models, backbone_option.model_type)
+            self.backbone_model = backbone_cls(architecture="unet", input_nc=input_nc, config=backbone_option)
+            self.conv_type = "DENSE"
```
Comment (nicolas-chaulet) [inline torch_points3d/applications/votenet.py:181]: change to `self.conv_type = self.backbone_model.conv_type`

File: torch_points3d/applications/votenet.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+                    data.x = data.ones.float()
+            self.input = Data(pos=data.pos, x=data.x, batch=data.batch).to(self.device)
+            self._num_batches = len(data.id_scan)
+
+    def forward(self, data):
+        self._set_input(data)
+
+        """Run forward pass. This will be called by both functions <optimize_parameters> and <test>."""
+        data_features = self.backbone_model.forward(self.input)
+        if self._backbone is None:
```
Comment (nicolas-chaulet) [inline torch_points3d/applications/votenet.py:252]: This is confusing. What you are testing here is wether we have a full Unet or if we shortcut it. It should be explicit. And for now as simple as `self._truncated_unet = backbone is None` in the init. You can remove the `_backbone` member, it does not tell much.

File: torch_points3d/applications/votenet.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        self.loss.backward()
+
+    def _extract_gt_center(self, data, outputs):
+        if self.is_dense_format:
+            gt_center = data.center_label[:, :, 0:3]
+        else:
+            gt_center = data.center_label[:, 0:3].view((self._num_batches, -1, 3))
+        outputs.assign_objects(gt_center, self.loss_params.near_threshold, self.loss_params.far_threshold)
+
+    def _convert_tensors_to_dense_format(self, data, outputs):
```
Comment (nicolas-chaulet) [inline torch_points3d/applications/votenet.py:283]: Please move this method to VoteNetResults. This is the class that should handle the specificity of output formatting. And also add a small test when you have done that please.

File: torch_points3d/core/data_transform/grid_transform.py [hunk 11]
```diff
@@ -16,7 +16,7 @@
 
 
 # Label will be the majority label in each voxel
-_INTEGER_LABEL_KEYS = ["y", "instance_labels"]
+INTEGER_LABEL_KEYS = ["y", "instance_labels"]
```
Comment (nicolas-chaulet) [inline torch_points3d/core/data_transform/grid_transform.py:19]: Leave it private

File: torch_points3d/modules/VoteNet/loss_helper.py [hunk 12]
```diff
@@ -47,11 +47,14 @@ def compute_vote_loss(input, output: VoteNetResults):
     #   non-object point has no GT vote mask = 0, object point has mask = 1
     # vote_label: Use gather to select B,num_seed,9 from B,num_point,9
     #   with inds in shape B,num_seed,9 and 9 = GT_VOTE_FACTOR * 3
-    seed_gt_votes_mask = torch.gather(input["vote_label_mask"], 1, seed_inds)
-    seed_inds_expand = seed_inds.view(batch_size, num_seed, 1).repeat(1, 1, 3 * GT_VOTE_FACTOR)
-
-    seed_gt_votes = torch.gather(input["vote_label"], 1, seed_inds_expand)
-    seed_gt_votes += output["seed_pos"].repeat(1, 1, 3)
+    if seed_inds.dim() == 1:
```
[the remark this answers is not in the document]
Comment (nicolas-chaulet) [reply torch_points3d/modules/VoteNet/loss_helper.py:50]: I think this hasn't been resolved

File: torch_points3d/modules/VoteNet/loss_helper.py [hunk 13]
```diff
@@ -115,11 +119,16 @@ def compute_box_and_sem_cls_loss(inputs, outputs, loss_params):
 
     # Compute center loss
     pred_center = outputs["center"]
-    gt_center = inputs["center_label"][:, :, 0:3]
+    if inputs["center_label"].dim() == 3:
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/loss_helper.py:122]: Please remove. All is dense at that point.

File: torch_points3d/modules/VoteNet/loss_helper.py [hunk 14]
```diff
@@ -115,11 +119,16 @@ def compute_box_and_sem_cls_loss(inputs, outputs, loss_params):
...
+        gt_center = inputs["center_label"][:, :, 0:3]
+    else:
+        gt_center = inputs["center_label"][:, 0:3].view((batch_size, -1, 3))
     dist1, ind1, dist2, _ = nn_distance(pred_center, gt_center)  # dist1: BxK, dist2: BxK2
     box_label_mask = inputs["box_label_mask"]
     objectness_label = outputs["objectness_label"].float()
     centroid_reg_loss1 = torch.sum(dist1 * objectness_label) / (torch.sum(objectness_label) + 1e-6)
+    if box_label_mask.dim() == 1:
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/loss_helper.py:130]: Please remove. All is dense at that point.

File: torch_points3d/modules/VoteNet/proposal_module.py [hunk 15]
```diff
@@ -58,7 +56,8 @@ def forward(self, data):
             VoteNetResults
         """
         if data.pos.dim() != 3:
-            raise ValueError("This method only supports dense convolutions for now")
+            raise Exception("data.pos doesn t have the correct dimension. Should be either 2 or 3")
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/proposal_module.py:59]: Please revert, it seems to only support dense for now.

File: torch_points3d/modules/VoteNet/voting_module.py (hunk 3 above)
Comment (nicolas-chaulet) [reply torch_points3d/modules/VoteNet/voting_module.py:30]: great

File: torch_points3d/applications/votenet.py (hunk 8 above)
Comment (tchaton) [reply torch_points3d/applications/votenet.py:181]: It doesn't work for some reason, I need to solve that.

**Comment by tchaton** [comment]
I have something really weird. I am trying to make a test for from torch_points3d.core.spatial_ops import RandomSamplerToDense.
However, python doesn't find it within the tests. I tried for 30 min and gave up...

File: torch_points3d/models/object_detection/votenet.py [hunk 16]
```diff
@@ -73,7 +73,8 @@ def set_input(self, data, device):
             input: a dictionary that contains the data itself and its metadata information.
         """
         # Forward through backbone model
-        self.input = data.to(device)
+        batch_size = data.pos.shape[0]
+        self.input = votenet_module.VoteNetResults.convert_tensors_to_dense_format(data.to(device), batch_size)
```
Comment (nicolas-chaulet) [inline torch_points3d/models/object_detection/votenet.py:77]: Not sure why this is here, could you explain?

File: torch_points3d/modules/VoteNet/votenet_results.py [hunk 17]
```diff
@@ -95,10 +95,25 @@ def from_logits(
...
-
+        if len(mean_size_arr) > 0:
+            data.size_residuals = size_residuals_normalized * mean_size_arr.unsqueeze(0).unsqueeze(0)
         sem_cls_scores = x_transposed[:, :, 5 + num_heading_bin * 2 + num_size_cluster * 4 :]  # Bxnum_proposalx10
         data.sem_cls_scores = sem_cls_scores
+
+        return data
+
+    @staticmethod
+    def convert_tensors_to_dense_format(data, batch_size):
```
Comment (nicolas-chaulet) [inline torch_points3d/modules/VoteNet/votenet_results.py:106]: Not really the correct place for that. Could you rename to `labels_to_dense` and move it to the VoteNet model class? It will be clearer like that

File: torch_points3d/models/object_detection/votenet.py (hunk 16 above)
Comment (nicolas-chaulet) [reply torch_points3d/models/object_detection/votenet.py:77]: Ok, got it, see other comment

File: torch_points3d/modules/VoteNet/votenet_results.py (hunk 17 above)
Comment (tchaton) [reply torch_points3d/modules/VoteNet/votenet_results.py]: It is used in the loss_helpers.py, so I moved it there. Make it cleaner.
